bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/ColorTemperature2.java

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright (c) 2010-2023 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.hue.internal.dto.clip2;
+
+import java.util.Objects;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * DTO for colour temperature of a light in CLIP 2.
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ */
+@NonNullByDefault
+public class ColorTemperature2 {
+    private @Nullable Integer mirek;
+    private @Nullable @SerializedName("mirek_schema") MirekSchema mirekSchema;
+
+    private float getReciprocal(float value) {
+        return Math.round(1000000f / value);
+    }
+
+    public @Nullable Integer getMirek() {
+        return mirek;
+    }
+
+    public void setMirek(float mirek) {
+        this.mirek = Math.round(mirek);
+    }
+
+    /**
+     * Convert the mirek value to a percentage value based on the passed MirekSchema argument.
+     *
+     * @param mirekSchema the reference MirekSchema.
+     * @return the percentage of the mirekSchema range.
+     */
+    public @Nullable Integer getPercent(MirekSchema mirekSchema) {
+        Integer mirek = this.mirek;
+        if (Objects.nonNull(mirek)) {
+            float min = mirekSchema.getMirekMinimum();
+            float max = mirekSchema.getMirekMaximum();
+            float percent = (100f * (mirek.floatValue() - min)) / (max - min);
+            return Math.round(Math.max(0, Math.min(100, percent)));
+        }
+        return null;
+    }
```

Comment:
Perhaps this logic should be kept in the handler to keep this DTO class clean of logic?

Reply:
Hmm. Can we please discuss this further?

My logic is that all the DTOs have 'regular' getter/setter methods to access the data fields in the form of primitive variables. PLUS also matching getter/setter methods to access the fields in form of Channel State variables.

I understand your pure logic that Channel States should not interfere in the concept of DTOs. But for my part, I am not comfortable to put them in the handler either. To be ultra pure, one might have 'middle-ware' classes between the handler and the DTO classes; but then I fear that @lolodomo may go even more bananas ;) about the extra classes..

Reply:
Can you elaborate why you don't feel comfortable having the API value to state conversion logic in the handler?

Reply:
> elaborate why you don't feel comfortable having the API value to state conversion logic in the handler

We have 26 DTO classes and one handler. The DTO classes are a hierarchical structure representing the detailed functionality of a device that supports/provides that respective DTO. And so I think the state conversion should be located within the same class that represents that detailed functionality. Otherwise we would move all the state conversion methods together into one lump within the handler, where there is no fine structure, and no direct overview of which method belongs to which functionality.

Reply:
Sorry for taking so long to respond, but I'm reopening this comment/discussion. By putting such logic into the DTO class, you are breaking the DTO Pattern. See for example https://www.baeldung.com/java-dto-pattern#common-mistakes.

The DTO should represent the contract and be used for serialization/deserialization.

Now, going through this specific class, here is my analysis and proposal:
- The class has only two members, `mirek` and `mirekSchema`.
- The logic in `getPercent` calculates percent based on `mirek` on the object itself and `mirekSchema` coming from the outside.
- This method is called from another DTO, `Resource`, in `getColorTemperaturePercentState`. Again `mirekSchema` is provided by the caller.
- This takes us back to `Clip2ThingHandler` which has a method called `mirekSchemaFrom`.
- This method calls `getMirekSchema` in `Resource`, which in turns calls `getMirekSchema` in `ColorTemperature2`.
- But also there is a fallback to another cached `ColorTemperature2` in case `mirekSchema` is missing in object at hand.

So in order to get the percent value (straight road without fallback):
- `Clip2ThingHandler.mirekSchemaFrom` -> `Resource.getMirekSchema` -> `ColorTemperature2.getMirekSchema`
- `Resource.getColorTemperaturePercentState` -> `ColorTemperature2.getPercent`

Logic is spread across three classes for this single operation.

I would propose to at least extract the logic into a utility class. The DTO's would be stripped of logic, and the handler would simply pass the data from the DTO's to the utility method in order for it to do its job. Unit test coverage would then also be trivial to add.

Basically this would mean that `getColorTemperaturePercentState` would be moved outside the DTO and would take two parameters, `mirekSchema` and `mirek`.

The handler code would not change much. Now:

```java
updateState(HueBindingConstants.CHANNEL_2_COLOR_TEMPERATURE,
    resource.getColorTemperaturePercentState(mirekSchemaFrom(resource)), fullUpdate);
```

With proposal:

```java
updateState(HueBindingConstants.CHANNEL_2_COLOR_TEMPERATURE,
    HueColorUtil.getColorTemperaturePercentState(mirekSchemaFrom(resource), resource.getColorTemperature()), fullUpdate);
```

Reply:
@andrewfg - perhaps this comment got lost?

Reply:
> perhaps this comment got lost?

It was at the back of my mind. I understand that you want the DTO classes to only contain 'setters' with a single argument, and getters (with a single return value). So yes, we could probably move the 'two argument setters' into a separate helper class. Before I do the refactoring can you please confirm that is what you mean?

Reply:
In general and by principle, yes. As long as translations are pretty much 1:1, some type conversions might be okay, this could be determined case by case. I'm simply opposing adding logic to the DTO classes like already mentioned in the comment above, which I believe is the best example I could find. With the proposal I made, I believe the handler/DTO entanglement/complexity would be resolved.